tests/typeparams/conversion_test.go

```diff
@@ -0,0 +1,250 @@
+		~float32 | ~float64 |
+		~uintptr
+}
+
+type numericConversion[srcType numeric, dstType numeric] struct {
+	src   srcType
+	want  dstType
+	quirk bool
+}
+
+func (tc numericConversion[srcType, dstType]) Run(t *testing.T) {
+	if tc.quirk && runtime.Compiler != "gopherjs" {
+		t.Skip("GopherJS-only test")
+	}
+
+	checkConversion(t, tc.src, dstType(tc.src), tc.want)
+}
+
+type complex interface {
+	~complex64 | ~complex128
+}
+
+type complexConversion[srcType complex, dstType complex] struct {
+	src  srcType
+	want dstType
+}
+
+func (tc complexConversion[srcType, dstType]) Run(t *testing.T) {
+	checkConversion(t, tc.src, dstType(tc.src), tc.want)
+}
+
+type stringLike interface {
+	// Ideally, we would test conversions from all integer types. unfortunately,
+	// that tips up the stringintconv check in `go vet` that is ran by `go test`
+	// by default. Unfortunately, there is no way to selectively suppress that
+	// check.
+	// ~int | ~int8 | ~int16 | ~int32 | ~int64 |
+	// ~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 |
+	// ~uintptr |
+	byte | rune |
+		~[]byte | ~[]rune | ~string
+}
+
+type stringConversion[srcType stringLike, dstType ~string] struct {
+	src  srcType
+	want dstType
+}
+
+func (tc stringConversion[srcType, dstType]) Run(t *testing.T) {
+	checkConversion(t, tc.src, dstType(tc.src), tc.want)
+}
+
+type boolConversion[srcType ~bool, dstType ~bool] struct {
+	src  srcType
+	want dstType
+}
+
+func (tc boolConversion[srcType, dstType]) Run(t *testing.T) {
+	checkConversion(t, tc.src, dstType(tc.src), tc.want)
+}
+
+// asString returns a string that reflects internal representation of the object.
+//
+// There is not specific guarantees about the string format, expect that if two
+// strings match, the two objects _almost certainly_ are deeply equal.
+func asString(o *js.Object) string {
+	f := js.Global.Get("Function").New("o", `
+	const seen = [];
+	// JSON.stringify can't deal with circular references, which GopherJS objects
+	// can have. So when the same object is seen more than once we replace it with
+	// a string stub.
+	const suppressCycles = (key, value) => {
+		if (typeof value !== 'object') {
+			return value;
+		}
+		const idx = seen.indexOf(value);
+		if (idx !== -1) {
+			return "[Cycle " + idx + "]"
+		}
+		seen.push(value);
+		return value;
+	}
+	return JSON.stringify(o, suppressCycles);
+	`)
+	return f.Invoke(o).String()
+}
+
+type interfaceConversion[srcType any] struct {
+	src srcType
+}
+
+func (tc interfaceConversion[srcType]) Run(t *testing.T) {
+	// All of the following expressions are semantically equivalent, but may be
+	// compiled by GopherJS differently, so we test all of them.
+	var got1 any
+	got1 = tc.src
+	var got2 any = tc.src
+	var got3 any = any(tc.src)
+
+	for i, got := range []any{got1, got2, got3} {
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
+			checkConversion(t, tc.src, got, tc.src)
+
+			concrete := got.(srcType) // Must not panic.
+			if runtime.Compiler == "gopherjs" {
```

Comment:
In light of the comment, Is there meant to be an `else` case that does use reflect.DeepEqual?